opt: GenerateLookupJoin panics on table with computed columns

Here's a reproduction:

```
exec-ddl
CREATE TABLE ab (a INT, b INT AS (a + 10) STORED, INDEX (a))
----

exec-ddl
CREATE TABLE cd (c INT, d INT)
----

opt
SELECT * FROM cd LEFT JOIN ab ON a = c AND d > 5
----
```

The error:

```
--- FAIL: TestNormRules (0.00s)
    --- FAIL: TestNormRules/__test (0.00s)
        opt_tester.go:506: testdata/rules/__test:42: runtime error: index out of range [-3]
            (1) assertion failure
            Wraps: (2) attached stack trace
              -- stack trace:
              | github.com/cockroachdb/cockroach/pkg/util/errorutil.ShouldCatch
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/util/errorutil/catch.go:29
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).Optimize.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:205
              | runtime.gopanic
              |         /usr/local/opt/go/libexec/src/runtime/panic.go:969
              | runtime.goPanicIndex
              |         /usr/local/opt/go/libexec/src/runtime/panic.go:88
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/testcat.(*Table).Column
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/testcat/test_catalog.go:656
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*CustomFuncs).findConstantFilterCols
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/select_funcs.go:417
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*CustomFuncs).computedColFilters
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/general_funcs.go:242
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*CustomFuncs).GenerateLookupJoins
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/join_funcs.go:227
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreLeftJoin
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:856
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreGroupMember
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.og.go:26
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*explorer).exploreGroup
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/explorer.go:182
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).optimizeGroup
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:464
              | github.com/cockroachdb/cockroach/pkg/sql/opt/xform.(*Optimizer).Optimize
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/xform/optimizer.go:225
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).optimizeExpr
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:1645
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).Optimize
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:981
              | github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester.(*OptTester).RunCommand
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/testutils/opttester/opt_tester.go:504
              | github.com/cockroachdb/cockroach/pkg/sql/opt/norm_test.TestNormRules.func1.1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/norm_test.go:45
              | github.com/cockroachdb/datadriven.runDirective.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:321
              | github.com/cockroachdb/datadriven.runDirective
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:326
              | github.com/cockroachdb/datadriven.runDirectiveOrSubTest
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:195
              | github.com/cockroachdb/datadriven.runTestInternal
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:168
              | github.com/cockroachdb/datadriven.RunTest
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:136
              | github.com/cockroachdb/cockroach/pkg/sql/opt/norm_test.TestNormRules.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/norm_test.go:42
              | github.com/cockroachdb/datadriven.Walk
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:413
              | github.com/cockroachdb/datadriven.Walk.func1
              |         /Users/marcus/go/src/github.com/cockroachdb/cockroach/vendor/github.com/cockroachdb/datadriven/datadriven.go:426
              | testing.tRunner
              |         /usr/local/opt/go/libexec/src/testing/testing.go:1123
              | runtime.goexit
              |         /usr/local/opt/go/libexec/src/runtime/asm_amd64.s:1374
            Wraps: (3) runtime error: index out of range [-3]
              | -- cause hidden behind barrier
              | runtime error: index out of range [-3]
              | (1) runtime error: index out of range [-3]
              | Error types: (1) runtime.boundsError
            Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *barriers.barrierError
```

--- a/pkg/sql/opt/xform/general_funcs.go
+++ b/pkg/sql/opt/xform/general_funcs.go
@@ -229,9 +229,9 @@ func (c *CustomFuncs) initIdxConstraintForIndex(
 // The values of both columns in that index are known, enabling a single value
 // constraint to be generated.
 func (c *CustomFuncs) computedColFilters(
-	tabID opt.TableID, requiredFilters, optionalFilters memo.FiltersExpr,
+	scanPrivate *memo.ScanPrivate, requiredFilters, optionalFilters memo.FiltersExpr,
 ) memo.FiltersExpr {
-	tabMeta := c.e.mem.Metadata().TableMeta(tabID)
+	tabMeta := c.e.mem.Metadata().TableMeta(scanPrivate.Table)
 	if len(tabMeta.ComputedCols) == 0 {
 		return nil
 	}
@@ -239,8 +239,8 @@ func (c *CustomFuncs) computedColFilters(
 	// Start with set of constant columns, as derived from the list of filter
 	// conditions.
 	constCols := make(map[opt.ColumnID]opt.ScalarExpr)
-	c.findConstantFilterCols(constCols, tabID, requiredFilters)
-	c.findConstantFilterCols(constCols, tabID, optionalFilters)
+	c.findConstantFilterCols(constCols, scanPrivate, requiredFilters)
+	c.findConstantFilterCols(constCols, scanPrivate, optionalFilters)
 	if len(constCols) == 0 {
 		// No constant values could be derived from filters, so assume that there
 		// are also no constant computed columns.

--- a/pkg/sql/opt/xform/general_funcs.go
+++ b/pkg/sql/opt/xform/general_funcs.go
@@ -11,11 +11,13 @@
 package xform
 
 import (
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/idxconstraint"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/memo"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/norm"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/partialidx"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/errors"
 )
 
@@ -259,4 +261,65 @@ func (c *CustomFuncs) computedColFilters(
 		}
 	}
 	return computedColFilters
+}
+
+// findConstantFilterCols adds to constFilterCols mappings from table column ID
+// to the constant value of that column. It does this by iterating over the
+// given lists of filters and finding expressions that constrain columns to a
+// single constant value. For example:
+//
+//   x = 5 AND y = 'foo'
+//
+// This would add a mapping from x => 5 and y => 'foo', which constants can
+// then be used to prove that dependent computed columns are also constant.
+func (c *CustomFuncs) findConstantFilterCols(
+	constFilterCols map[opt.ColumnID]opt.ScalarExpr,
+	scanPrivate *memo.ScanPrivate,
+	filters memo.FiltersExpr,
+) {
+	tab := c.e.mem.Metadata().Table(scanPrivate.Table)
+	for i := range filters {
+		// If filter constraints are not tight, then no way to derive constant
+		// values.
+		props := filters[i].ScalarProps()
+		if !props.TightConstraints {
+			continue
+		}
+
+		// Iterate over constraint conjuncts with a single column and single
+		// span having a single key.
+		for i, n := 0, props.Constraints.Length(); i < n; i++ {
+			cons := props.Constraints.Constraint(i)
+			if cons.Columns.Count() != 1 || cons.Spans.Count() != 1 {
+				continue
+			}
+
+			// Skip columns that aren't in the scanned table.
+			colID := cons.Columns.Get(0).ID()
+			if !scanPrivate.Cols.Contains(colID) {
+				continue
+			}
+
+			// Skip columns with a data type that uses a composite key encoding.
+			// Each of these data types can have multiple distinct values that
+			// compare equal. For example, 0 == -0 for the FLOAT data type. It's
+			// not safe to treat these as constant inputs to computed columns,
+			// since the computed expression may differentiate between the
+			// different forms of the same value.
+			colTyp := tab.Column(scanPrivate.Table.ColumnOrdinal(colID)).DatumType()
+			if colinfo.HasCompositeKeyEncoding(colTyp) {
+				continue
+			}
+
+			span := cons.Spans.Get(0)
+			if !span.HasSingleKey(c.e.evalCtx) {
+				continue
+			}
+
+			datum := span.StartKey().Value(0)
+			if datum != tree.DNull {
+				constFilterCols[colID] = c.e.f.ConstructConstVal(datum, colTyp)
+			}
+		}
+	}
 }

--- a/pkg/sql/opt/xform/join_funcs.go
+++ b/pkg/sql/opt/xform/join_funcs.go
@@ -224,7 +224,7 @@ func (c *CustomFuncs) GenerateLookupJoins(
 	// Generate implicit filters from CHECK constraints and computed columns as
 	// optional filters to help generate lookup join keys.
 	optionalFilters := c.checkConstraintFilters(scanPrivate.Table)
-	computedColFilters := c.computedColFilters(scanPrivate.Table, on, optionalFilters)
+	computedColFilters := c.computedColFilters(scanPrivate, on, optionalFilters)
 	optionalFilters = append(optionalFilters, computedColFilters...)
 
 	var pkCols opt.ColList
@@ -505,7 +505,7 @@ func (c *CustomFuncs) GenerateInvertedJoins(
 				// using them here would result in a reduced set of optional
 				// filters.
 				optionalFilters = c.checkConstraintFilters(scanPrivate.Table)
-				computedColFilters := c.computedColFilters(scanPrivate.Table, on, optionalFilters)
+				computedColFilters := c.computedColFilters(scanPrivate, on, optionalFilters)
 				optionalFilters = append(optionalFilters, computedColFilters...)
 
 				eqColsAndOptionalFiltersCalculated = true

--- a/pkg/sql/opt/xform/select_funcs.go
+++ b/pkg/sql/opt/xform/select_funcs.go
@@ -13,7 +13,6 @@ package xform
 import (
 	"sort"
 
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
 	"github.com/cockroachdb/cockroach/pkg/sql/inverted"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/cat"
@@ -368,67 +367,6 @@ func (c *CustomFuncs) GenerateConstrainedScans(
 
 		sb.build(grp)
 	})
-}
-
-// findConstantFilterCols adds to constFilterCols mappings from table column ID
-// to the constant value of that column. It does this by iterating over the
-// given lists of filters and finding expressions that constrain columns to a
-// single constant value. For example:
-//
-//   x = 5 AND y = 'foo'
-//
-// This would add a mapping from x => 5 and y => 'foo', which constants can
-// then be used to prove that dependent computed columns are also constant.
-func (c *CustomFuncs) findConstantFilterCols(
-	constFilterCols map[opt.ColumnID]opt.ScalarExpr,
-	scanPrivate *memo.ScanPrivate,
-	filters memo.FiltersExpr,
-) {
-	tab := c.e.mem.Metadata().Table(scanPrivate.Table)
-	for i := range filters {
-		// If filter constraints are not tight, then no way to derive constant
-		// values.
-		props := filters[i].ScalarProps()
-		if !props.TightConstraints {
-			continue
-		}
-
-		// Iterate over constraint conjuncts with a single column and single
-		// span having a single key.
-		for i, n := 0, props.Constraints.Length(); i < n; i++ {
-			cons := props.Constraints.Constraint(i)
-			if cons.Columns.Count() != 1 || cons.Spans.Count() != 1 {
-				continue
-			}
-
-			// Skip columns that aren't in the scanned table.
-			colID := cons.Columns.Get(0).ID()
-			if !scanPrivate.Cols.Contains(colID) {
-				continue
-			}
-
-			// Skip columns with a data type that uses a composite key encoding.
-			// Each of these data types can have multiple distinct values that
-			// compare equal. For example, 0 == -0 for the FLOAT data type. It's
-			// not safe to treat these as constant inputs to computed columns,
-			// since the computed expression may differentiate between the
-			// different forms of the same value.
-			colTyp := tab.Column(scanPrivate.Table.ColumnOrdinal(colID)).DatumType()
-			if colinfo.HasCompositeKeyEncoding(colTyp) {
-				continue
-			}
-
-			span := cons.Spans.Get(0)
-			if !span.HasSingleKey(c.e.evalCtx) {
-				continue
-			}
-
-			datum := span.StartKey().Value(0)
-			if datum != tree.DNull {
-				constFilterCols[colID] = c.e.f.ConstructConstVal(datum, colTyp)
-			}
-		}
-	}
 }
 
 // tryFoldComputedCol tries to reduce the computed column with the given column

--- a/pkg/sql/opt/xform/select_funcs.go
+++ b/pkg/sql/opt/xform/select_funcs.go
@@ -209,7 +209,7 @@ func (c *CustomFuncs) GenerateConstrainedScans(
 	// Generate implicit filters from constraints and computed columns as
 	// optional filters to help constrain an index scan.
 	optionalFilters := c.checkConstraintFilters(scanPrivate.Table)
-	computedColFilters := c.computedColFilters(scanPrivate.Table, explicitFilters, optionalFilters)
+	computedColFilters := c.computedColFilters(scanPrivate, explicitFilters, optionalFilters)
 	optionalFilters = append(optionalFilters, computedColFilters...)
 
 	filterColumns := c.FilterOuterCols(explicitFilters)
@@ -380,9 +380,11 @@ func (c *CustomFuncs) GenerateConstrainedScans(
 // This would add a mapping from x => 5 and y => 'foo', which constants can
 // then be used to prove that dependent computed columns are also constant.
 func (c *CustomFuncs) findConstantFilterCols(
-	constFilterCols map[opt.ColumnID]opt.ScalarExpr, tabID opt.TableID, filters memo.FiltersExpr,
+	constFilterCols map[opt.ColumnID]opt.ScalarExpr,
+	scanPrivate *memo.ScanPrivate,
+	filters memo.FiltersExpr,
 ) {
-	tab := c.e.mem.Metadata().Table(tabID)
+	tab := c.e.mem.Metadata().Table(scanPrivate.Table)
 	for i := range filters {
 		// If filter constraints are not tight, then no way to derive constant
 		// values.
@@ -399,14 +401,19 @@ func (c *CustomFuncs) findConstantFilterCols(
 				continue
 			}
 
+			// Skip columns that aren't in the scanned table.
+			colID := cons.Columns.Get(0).ID()
+			if !scanPrivate.Cols.Contains(colID) {
+				continue
+			}
+
 			// Skip columns with a data type that uses a composite key encoding.
 			// Each of these data types can have multiple distinct values that
 			// compare equal. For example, 0 == -0 for the FLOAT data type. It's
 			// not safe to treat these as constant inputs to computed columns,
 			// since the computed expression may differentiate between the
 			// different forms of the same value.
-			colID := cons.Columns.Get(0).ID()
-			colTyp := tab.Column(tabID.ColumnOrdinal(colID)).DatumType()
+			colTyp := tab.Column(scanPrivate.Table.ColumnOrdinal(colID)).DatumType()
 			if colinfo.HasCompositeKeyEncoding(colTyp) {
 				continue
 			}
@@ -723,7 +730,7 @@ func (c *CustomFuncs) GenerateInvertedIndexScans(
 	// Generate implicit filters from constraints and computed columns as
 	// optional filters to help constrain an index scan.
 	optionalFilters := c.checkConstraintFilters(scanPrivate.Table)
-	computedColFilters := c.computedColFilters(scanPrivate.Table, filters, optionalFilters)
+	computedColFilters := c.computedColFilters(scanPrivate, filters, optionalFilters)
 	optionalFilters = append(optionalFilters, computedColFilters...)
 
 	// Iterate over all inverted indexes.

--- a/pkg/sql/opt/xform/testdata/rules/join
+++ b/pkg/sql/opt/xform/testdata/rules/join
@@ -3174,6 +3174,35 @@ anti-join (hash)
  └── filters
       └── column1:1 = y:3 [outer=(1,3), constraints=(/1: (/NULL - ]; /3: (/NULL - ]), fd=(1)==(3), (3)==(1)]
 
+# Regression test for #59738. findConstantFilterCols should not panic when it
+# finds constant filters for columns that are not in the index of the lookup
+# table.
+exec-ddl
+CREATE TABLE t59738_ab (a INT, b INT AS (a + 10) STORED, INDEX (a))
+----
+
+exec-ddl
+CREATE TABLE t59738_cd (c INT, d INT)
+----
+
+opt expect=GenerateLookupJoins
+SELECT * FROM t59738_cd LEFT LOOKUP JOIN t59738_ab ON a = c AND d > 5
+----
+left-join (lookup t59738_ab)
+ ├── columns: c:1 d:2 a:5 b:6
+ ├── key columns: [7] = [7]
+ ├── lookup columns are key
+ ├── left-join (lookup t59738_ab@secondary)
+ │    ├── columns: c:1 d:2 a:5 t59738_ab.rowid:7
+ │    ├── flags: force lookup join (into right side)
+ │    ├── key columns: [1] = [5]
+ │    ├── fd: (7)-->(5)
+ │    ├── scan t59738_cd
+ │    │    └── columns: c:1 d:2
+ │    └── filters
+ │         └── d:2 > 5 [outer=(2), constraints=(/2: [/6 - ]; tight)]
+ └── filters (true)
+
 # --------------------------------------------------
 # GenerateLookupJoinsWithFilter
 # --------------------------------------------------

